File: 백준/Silver/2343. 기타 레슨/기타 레슨.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	static int N, M, arr[], ans;
	static int start, end; 
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		N=Integer.parseInt(st.nextToken());
		M=Integer.parseInt(st.nextToken());
		start=0;
		arr=new int[N];
		st=new StringTokenizer(br.readLine());
		for(int i=0;i<N;i++) {
			arr[i]=Integer.parseInt(st.nextToken());
			start=Math.max(arr[i], start);
			end+=arr[i];
		}
		ans=Integer.MAX_VALUE;
		//start는 블루레이 중 가장 작은 값
		//end는 전부 합
		//mid는 블루레이의 최종 크기
		int mid=0;
		while(start<=end) {
			mid=(start+end)/2;
			int cnt=1;
			int sum=0;
			for(int i=0;i<N;i++) {
				if(sum+arr[i]>mid) {
					cnt++;
					sum=arr[i];
				}
				else {
					sum+=arr[i];
				}
			}
			if(cnt<=M) {
				end=mid-1;
			}
			else {
				//분 늘려야 함
				start=mid+1;
			}
		}
		System.out.println(start);
		
	}

}

File: 백준/Silver/2343. 기타 레슨/기타 레슨.cc
#include <iostream>
#include <algorithm>
using namespace std;
int n, m;
int arr[100001];
int st, en;
int main()
{
    cin >> n >> m;
    for (int i = 0; i < n; i++)
    {
        cin >> arr[i];
        en += arr[i];
        st = max(arr[i], st);
    }
    // en -> 강의 길이의 합
    // st -> 가장 긴 강의
    while (st <= en)
    {
        int mid = (st + en) / 2; // 한 블루레이의 크기
        int sum = 0;
        int cnt = 1;
        for (int i = 0; i < n; i++)
        {
            // 블루레이의 크기가 mid보다 커지면 cnt 증가
            if (sum + arr[i] <= mid)
            {
                sum += arr[i];
            }
            else
            {
                cnt++;
                sum = arr[i];
            }
        }
        if (cnt <= m)
        {
            // 블루레이 크기 줄임
            en = mid - 1;
        }
        else
        {
            // 블루레이 크기 늘림
            st = mid + 1;
        }
    }
    cout << st;
}